packages/SwingSet/src/limits/calculate.js

```diff
@@ -0,0 +1,193 @@
+// @ts-check
+
+/** @typedef {import('./types').Budget} Budget */
+/** @typedef {import('./types').MemoryCostModel} MemoryCostModel */
+
+const { details: X } = assert;
+// @ts-expect-error verbatim is not defined
+const v = assert.verbatim || assert.quote;
+
+export const BIGINT_WORD_VALUE = 256n ** 8n;
+
+/**
+ * @param {MemoryCostModel} costModel
+ */
+export const makeCalculateCost = costModel => {
+  assert.typeof(
+    costModel,
+    'object',
+    X`costModel ${costModel} must be an object`,
+  );
+  const {
+    baseCost,
+    bigintPerWordCost,
+    stringPerCharacterCost,
+    objectPerPropertyCost,
+  } = costModel;
+  assert.typeof(baseCost, 'bigint', X`baseCost ${baseCost} must be a bigint`);
+  assert.typeof(
+    bigintPerWordCost,
+    'bigint',
+    X`bigintPerWordCost ${bigintPerWordCost} must be a bigint`,
+  );
+  assert.typeof(
+    stringPerCharacterCost,
+    'bigint',
+    X`stringPerCharacterCost ${stringPerCharacterCost} must be a bigint`,
+  );
+  assert.typeof(
+    objectPerPropertyCost,
+    'bigint',
+    X`objectPerPropertyCost ${objectPerPropertyCost} must be a bigint`,
+  );
+
+  let totalCost = 0n;
+  const getTotalCost = () => totalCost;
+
+  /**
+   * @param {unknown} value
+   */
+  const shallowCost = value => {
+    let cost = baseCost;
+    if (Object(value) === value) {
+      /**
+       * Either a function or an object with properties.
+       *
+       * @type {{ [x: PropertyKey]: PropertyDescriptor }}
+       */
+      const descs = Object.getOwnPropertyDescriptors(value);
+      const keys = Reflect.ownKeys(descs);
+      cost += BigInt(keys.length) * objectPerPropertyCost;
+      totalCost += cost;
+      return cost;
+    }
+    // We have a primitive.
+    switch (typeof value) {
+      case 'string': {
+        // The size of the string, in approximate bytes.
+        cost += BigInt(value.length) * stringPerCharacterCost;
+        break;
+      }
+      case 'bigint': {
+        // Compute the number of approximate bytes in the bigint.
+        let remaining = value;
+        if (remaining < 0n) {
+          remaining = -remaining;
+        }
+        // Count number of bigint chunk and multiply out.
+        while (remaining > 0n) {
+          remaining /= BIGINT_WORD_VALUE;
```

Comment:
Dividing is expensive. How about pre-computing the max?

Reply:
Oh... This is a cost, not a threshold. Still... A table of word-sizes could be pre-computed.

Reply:
I can make it way more efficient by stringifying the bigint then counting the digits.  So I will.